### ISSUE TITLE
Weird behavior of subprocess with Pyinstaller compiled on ubuntu and run on SuSe 12 SP4

### ISSUE THREAD

**Issue by pranky89** [pyinstaller/pyinstaller#5429]
I am noticing some weird behavior of subprocess compiled using pyinstaller on Ubuntu 16 which works fine on ubuntu 16 but fails on SuSe12 SP4. Was wondering if someone can shed some light and tell me what trivial info Am i missing?

Basically, I have a simple script :

```
import os, sys, subprocess
def get_all_outputs(cmd):
    proc = subprocess.Popen(cmd,
                            stdout=subprocess.PIPE,
                            stderr=subprocess.PIPE,
                            shell=True,
                            universal_newlines=True)
    std_out, std_err = proc.communicate()
    return proc.returncode, std_out, std_err
cmd = raw_input('CMD:')
print get_all_outputs(cmd)

```

This works fine as normal python file or pyinstaller file on Ubuntu 16...

```
root@ubuntu16:~/cert/dist# ./retcode
CMD:openssl x509 -in /root/cert_new/mycert.cer
```

`(0, '-----BEGIN CERTIFICATE-----\nMIIFmTCCBIGgAwIBAgITQwAAAGqQd2QfUVAHwQABAAAAajANBgkqhkiG9w0BAQsF\nADBGMRUwEwYKCZImiZPyLGQBGRYFbG9jYWwxFjAUBgoJkiaJk/IsZAEZFgZhdmFt\nYXIxFTATBgNVBAMTDGF2YW1hci1EQy1DQTAeFw0yMDEyMjkxMjAyNDhaFw0yMjEy\nMjkxMjAyNDhaMHYxCzAJBgNVBAYTAkFVMQwwCgYDVQQIEwNOU1cxDzANBgNVBAcT\nBlN5ZG5leTERMA8GA1UEChMIRGVsbCBFTUMxEDAOBgNVBAsTB1N1cHBvcnQxIzAh\nBgNVBAMTGm5lbzEtc3lWwMBMA0GCSqGSIb3DQEBCwUAA4IBAQBjH2ubfVxCC42LVURTZUog/vJZ\nctAEBDUW3VaeRCMWD3dvB0loc0llGaXQVafh0Q2cW8Uy0qMexPcUUwp8OjbtwcBo\n3TkEApBABgX/JC9P+BXCK3NiYze1SAjsgcdeZaS0t3HLlgwc8vZSotXco+mwZM9S\nTtrU1RqU4kkqhR5+wjPT8ffLFyZNBCdDKUOF3wxsr/0uUpfm9Bnt3DahoN4dwHvI\nOvi1DSV6ob84VXKT3ehMqt27ZW5dtLQdpzINADHDHlitTAUAO+CdO3LltqobQbf8\niK8fmnmSWHVF8vA3mmIfANLILZ6XKASgo2D2RU0jPjbkWi3nPY+2aRPGS1wJ\n-----END CERTIFICATE-----\n', '')`

But when I scp the compiled file to SuSe12SP4, I never get the output and instead it keeps throwing error code 127 along with a funny message like below:

```
neo_suse12sp4:~/pp # ./retcode
CMD:/usr/local/ssl/bin/openssl x509 -in /root/pp/mycert.cer

(127, '', '/bin/sh: /tmp/_MEIOqKDWs/libreadline.so.6: no version information available (required by /bin/sh)\n/usr/local/ssl/bin/openssl: relocation error: /usr/local/ssl/bin/openssl: symbol i2d_DHxparams, version OPENSSL_1.0.0 not defined in file libcrypto.so.1.0.0 with link time reference\n')
```

I even tried using full path to openssl and the certificate and it simply never works.

```
neo_suse12sp4:~/pp # which openssl
/usr/local/ssl/bin/openssl
```
I would really appreciate if someone can please assist me here. I'll be really thankful to you. Google and existing articles on SO seems to be no match to this problem.

**Comment by rokm** [comment]
> /usr/local/ssl/bin/openssl: relocation error: /usr/local/ssl/bin/openssl: symbol i2d_DHxparams, version OPENSSL_1.0.0 not defined in file libcrypto.so.1.0.0 with link time reference'

This looks like the `openssl` executable found on your SuSe 12 SP4 system is incompatible with the `libcrypto.so` library that was collected from Ubuntu 16 build system (and is now overriding the system one in the frozen application and its subprocesses).

You'll probably need to modify the LD_LIBRARY_PATH for your subprocess to make the system's libraries take precedence over the bundled ones. [There's a topic in the manual about that.](https://pyinstaller.readthedocs.io/en/stable/runtime-information.html?ld-library-path-libpath-considerations#ld-library-path-libpath-considerations)

(If the system does not have LD_LIBRARY_PATH explicitly set, then you will probably need to reset it for the subprocess as well, instead of using the PyInstaller-set one).

**Comment by pranky89** [comment]
Thank @rokm for suggestion. I added the following code at the top of my code and it still fails with same error.
Can you be please specific on what I should be editing in my simple code to make it work ?

```
env = dict(os.environ)  # make a copy of the environment
lp_key = 'LD_LIBRARY_PATH'  # for GNU/Linux and *BSD.
lp_orig = env.get(lp_key + '_ORIG')
if lp_orig is not None:
    env[lp_key] = lp_orig  # restore the original, unmodified value
else:
    # This happens when LD_LIBRARY_PATH was not set.
    # Remove the env var as a last resort:
    env.pop(lp_key, None)
```

**Comment by rokm** [comment]
Did you pass `env` to your `Popen()` call?

**Comment by pranky89** [comment]
Thank you so so much @rokm for your great comment, the small aforementioned code indeed started working after that addition but here is the problem. When I added the same code to my other program where I have a function subprocess_cmd, it is still giving output as '127'. I am using it like this : (its a flask app)

I am also seeing the line : sh: /tmp/_MEIRm55bP/libreadline.so.6: no version information available (required by sh)

I would greatly appreciate your assistance.....Thanks a lot for your assistance already.

```
def run_command(cmd): #returns the output of a program
    proc = subprocess.Popen(cmd,stdout=subprocess.PIPE, stderr=subprocess.PIPE, stdin=subprocess.PIPE, shell=True, universal_newlines=True, env=env)
    std_out, std_err = proc.communicate()
    return std_out

def check_certficate(): #checkcert.sh is external script that runs openssl command and returns exit code
    cmd = 'bash checkcert.sh abc.cer'
    j = run_command(cmd)
   print j
   if j not 0:
        print 'problem found'

if __name__ == '__main__':
    env = dict(os.environ)  # make a copy of the environment
    lp_key = 'LD_LIBRARY_PATH'  # for GNU/Linux and *BSD.
    lp_orig = env.get(lp_key + '_ORIG')
    logger.info('Reviewed environment information...')
    if lp_orig is not None:
        env[lp_key] = lp_orig  # restore the original, unmodified value
    else:
        # This happens when LD_LIBRARY_PATH was not set.
        # Remove the env var as a last resort:
        env.pop(lp_key, None)

    app.jinja_env.cache = {}
    app.run(host='0.0.0.0', port='17000', debug=True, threaded=True)   
```

**Comment by pranky89** [comment]
Thank you so so much @rokm for your great comment, the small aforementioned code indeed started working after that addition.
The only thing that needs to be fixed is this line which is coming again and again :
`
sh: /tmp/_MEIRm55bP/libreadline.so.6: no version information available (required by sh)`

How can I get rid of it?

**Comment by rokm** [comment]
>  sh: /tmp/_MEIRm55bP/libreadline.so.6: no version information available (required by sh)

Is this actually causing any issues?

Similarly to before, it is again caused by the `libreadline.so` that was collected from your Ubuntu system not being entirely compatible with the one that's used on the target system (which the `sh` interpreter was originally linked to; and the interpreter is spawned because you are using `shell=True` option in `Popen()`).

I don't think I have seen the lack of version information causing any real issues before (aside from the warning message). But if you want to get rid of it, you could try omitting the `shell` option (not sure if you really need it or not). If that's not an option, then you'll need to find a way to make the system version of `libreadline.so` take precedence (I'm actually a bit surprised that resetting the LD_LIBRARY_PATH did not take care of that, too...)

**Comment by pranky89** [comment]
You are right, I don't really need shell=True, therefore, i'll go ahead and remove it from my master code as well.
Also, after using your suggestion of env, both shell=true and without are not throwing that message in test code but my code is still throwing it.
What i'll do is, i'll make some changes in my code such as removing shell=true from my actual code and then will give it a shot, i am confident that it will work.
TLDR: Thank you for your support. It was really really great that you pointed it out. Happy new year and have a great day.
